pmd/pm_echo.c

```diff
@@ -0,0 +1,800 @@
+	else {
+		pme->retry.retry_already_counted = true;
+		pme->retry.retry_count++;
+	}
+	/* check down or up context */
+	if (pme->retry.retry_count < counter) {
+		THREAD_OFF(pme->t_echo_tmo);
+		return true;
+	}
+	/* else fall on timeout */
+	return false;
+}
+
+int pm_echo_tmo(struct thread *thread)
+{
+	struct pm_echo *pme = THREAD_ARG(thread);
+	struct pm_session *pm = (struct pm_session *)pme->back_ptr;
+
+	if (pme->echofd < 0)
+		return 0;
+	/* else fall on timeout */
+	if (pme->oper_receive) {
+		zlog_info("PMD: packet already received. cancel tmo");
+		return 0;
+	}
+	pme->stats_rx_timeout++;
+	if (pm_check_retries(pme, pme->retries_down, false))
+		return 0;
+	/* reset pme retries context */
+	pme->retry.retry_count = 0;
+	pme->retry.retry_down_in_progress = false;
+	pme->retry.retry_up_in_progress = false;
+	pm_echo_trigger_down_event(pm);
+	return 0;
+}
+
+int pm_echo_receive(struct thread *thread)
+{
+	int fd = THREAD_FD(thread);
+	struct pm_echo *pme = THREAD_ARG(thread);
+	struct pm_session *pm = pme->back_ptr;
+	struct sockaddr from;
+	struct icmphdr *icmp;
+	struct icmp6_hdr *icmp6;
+	socklen_t fromlen = sizeof(from);
+	int hlen, ret = 0;
+	struct iphdr *ip;
+	char buf[INET6_ADDRSTRLEN];
+
+	if ((sockunion_family(&pme->peer) == AF_INET
+	     && pme->echofd < 0) ||
+	    (sockunion_family(&pme->peer) == AF_INET6
+	     && pme->echofd_rx_ipv6 < 0)) {
+		/* context flushed; prevent from read */
+		fd = -1;
+	}
+	if (fd >= 0)
+		thread_add_read(master, pm_echo_receive, pme,
+				fd, &pme->t_echo_receive);
+	if (!pme->rx_buf)
+		pme->rx_buf = XCALLOC(MTYPE_PM_PACKET, pme->packet_size);
+	else
+		memset(pme->rx_buf, 0, pme->packet_size);
+	icmp = (struct icmphdr *)pme->rx_buf;
+	ret = recvfrom(fd, pme->rx_buf,
+		       pme->packet_size, 0, &from, &fromlen);
+	if (ret < 0 || fd < 0) {
+		zlog_err("PMD: error when receiving ICMP echo.");
+		return 0;
+	}
+	monotime(&pme->end);
+	if (sockunion_family(&pme->peer) == AF_INET) {
+		ip = (struct iphdr *)pme->rx_buf;
+		hlen = ip->ihl << 2;
+		/* check that destination address matches
+		 * our local address configured
+		 */
+		if (ip->saddr != pme->peer.sin.sin_addr.s_addr) {
+			if (pm_debug_echo) {
+				char buf2[INET6_ADDRSTRLEN];
+				struct in_addr saddr;
+
+				saddr.s_addr = ip->saddr;
+				inet_ntop(AF_INET, &saddr, buf, sizeof(buf));
+				inet_ntop(AF_INET, &pme->peer.sin.sin_addr,
+					  buf2, sizeof(buf2));
+				zlog_err("PMD: wrong src address %s, expected %s. retrying",
+					 buf, buf2);
+			}
+			return 0;
+		}
+		if (sockunion_family(&pm->key.local) == AF_INET &&
+		    ip->daddr != pm->key.local.sin.sin_addr.s_addr) {
+			if (pm_debug_echo) {
+				char buf2[INET6_ADDRSTRLEN];
+				struct in_addr daddr;
+
+				daddr.s_addr = ip->daddr;
+				inet_ntop(AF_INET, &daddr, buf, sizeof(buf));
+				inet_ntop(AF_INET, &pm->key.local.sin.sin_addr,
+					  buf2, sizeof(buf2));
+				zlog_err("PMD: wrong dst address %s, expected %s. retrying",
```

Comment:
Better use FRR formatting here:
`%pFX` for `struct prefix` and `%pI4`/`%pI6` for IPv4/IPv6 addresses.

More info:
http://docs.frrouting.org/projects/dev-guide/en/latest/logging.html#extensions